### PR TITLE
MaxPool 8 reference implementation

### PR DESCRIPTION
**Pull request by tomdol**
### Details:
 - The `kernel_position` variable represented as the `Coord<T>` object tracks the top-left corner of the kernel. If the kernel overlaps with the padding area, the coordinates of the top-left corner are negative.
 - The `kernel_offset` is used to to track the currently considered/processed pixel under a kernel. The offsets is always a positive number and represents a distance from the top-left corner of the kernel. For non-dilated kernel the offsets for each kernel axis are ascending ranges `[0, 1, 2, ..., kernel_size)`. For dilated kernels the ranges are `[0, dilation, 2*dilation, 3*dilation, ...)`
 - Those 2 variables are helpers used to calculate the current pixel index in the flattened representation of the input tensor
 - They are also used to check if a given element under a kernel position belongs to the padding area.

image

### Tickets:
 - *54659*

### CODE REVIEW

File: ngraph/core/src/op/max_pool.cpp [hunk 1]
```diff
@@ -227,3 +319,54 @@ shared_ptr<Node> op::v8::MaxPool::clone_with_new_inputs(const OutputVector& new_
                                     m_axis,
                                     m_pads_value);
 }
+
+bool op::v8::MaxPool::has_evaluate() const {
+    NGRAPH_OP_SCOPE(v8_MaxPool_has_evaluate);
+    switch (get_input_element_type(0)) {
```
Comment (tomdol) [inline ngraph/core/src/op/max_pool.cpp:325]: Should I enable it for other data types as well? In particular I'm thinking about int8 and uint8

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp [hunk 2]
```diff
@@ -110,6 +110,340 @@ void max_pool(const T* arg,
...
+        }
+    }
+
+    Coord(std::initializer_list<T>&& values) : std::vector<T>{std::move(values)} {}
+};
+
+bool elem_in_padding_area(const Coord<int>& kernel_position,
+                          const Coord<size_t>& kernel_offset,
+                          const Shape& data_shape) {
+    for (size_t dim = 0; dim < data_shape.size() - 2; ++dim) {
```
Comment (mitruska) [inline ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:152]: Do we really need a loop for such check?

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp [hunk 3]
```diff
@@ -110,6 +110,340 @@ void max_pool(const T* arg,
...
+                 "\npads_end=",
+                 pads_end);
+}
+
+template <typename T>
+struct Coord : public std::vector<T> {
+    Coord(const Shape& pads_begin) {
+        std::vector<T>::reserve(pads_begin.size());
+        for (const auto axis_padding : pads_begin) {
+            std::vector<T>::push_back(0 - axis_padding);
```
Comment (mitruska) [inline ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:142]: I have to familiarize myself with this new `Coord` struct to understand the rest of the code, not sure I like it 🤓

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp [hunk 4]
```diff
@@ -110,6 +110,340 @@ void max_pool(const T* arg,
...
+void max_pool(const Values_t* data,
+              Values_t* values,
+              Indices_t* indices,
+              const Shape& data_shape,
+              const Shape& out_shape,
+              const Shape& kernel,
+              const Strides& strides,
+              const Strides& dilations,
+              const Shape& pads_begin,
+              const Shape& pads_end,
```
[the remark this answers is not in the document]
Comment (tomdol) [reply ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:370]: This is a kind of convention across OV to use the `Shape` objects to store padding configuration.

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp (hunk 2 above)
Comment (tomdol) [reply ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:152]: The loop makes this function work for 2D and 3D. Otherwise I would have to hardcode manual check separately for 2D and 3D and this code would grow.

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp (hunk 3 above)
Comment (tomdol) [reply ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:142]: `Coord` is just like `Coordinate` except it can store different data types for the underlying dimensions. I need signed integers for kernel positions and `Coordinate` only allows for `size_t`.

I can make the `Coordinate` class a template with `size_t` as default and use it here instead if that makes sense.

Comment (tomdol) [reply ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:142]: I've done an experiment and I can't make `Coordinate` a template class without huge amount of changes in the whole code base. I'd like to stick to this helper struct unless there are some very strong arguments to drop it.

**Comment by tomdol** [comment]
@ilyachur since there's no plugin implementation yet and there's nothing to compare against, should this PR contain single layer tests or should they be delivered with any plugin implementation (CPU first I assume)?

**Comment by ilyachur** [comment]
> 
> 
> @ilyachur since there's no plugin implementation yet and there's nothing to compare against, should this PR contain single layer tests or should they be delivered with any plugin implementation (CPU first I assume)?

It is strange to have tests which are disabled for all components 😄 I think we can add SLT tests in the separate PR.

**Review by mbencer — APPROVED** [review]

**Review by tsavina — APPROVED** [review]

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp [hunk 5]
```diff
@@ -110,6 +110,340 @@ void max_pool(const T* arg,
...
+                 "\nkernel_dilations=",
+                 kernel_dilations,
+                 "\npads_begin=",
+                 pads_begin,
+                 "\npads_end=",
+                 pads_end);
+}
+
+template <typename T>
+struct Coord : public std::vector<T> {
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:138]: @tomdol I think we need to provide more documentation for this code, can you do it in the separate PR?

File: ngraph/core/src/op/max_pool.cpp [hunk 6]
```diff
@@ -223,3 +319,56 @@ shared_ptr<Node> op::v8::MaxPool::clone_with_new_inputs(const OutputVector& new_
...
+    case ngraph::element::f32:
+        return true;
+    default:
+        break;
+    }
+    return false;
+}
+
+bool op::v8::MaxPool::evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs) const {
+    NGRAPH_OP_SCOPE(v8_MaxPool_evaluate);
```
Comment (ilyachur) [inline ngraph/core/src/op/max_pool.cpp:342]: Do we really need to have evaluate for this operation?

**Review by ilyachur — APPROVED** [review]

File: ngraph/core/include/ngraph/shape.hpp [hunk 7]
```diff
@@ -57,6 +57,18 @@ size_t shape_size(const SHAPE_TYPE& shape) {
     return size;
 }
 
+template <typename ForwardIt>
+size_t shape_size(ForwardIt start_dim, const ForwardIt end_dim) {
+    static_assert(std::is_arithmetic<typename std::iterator_traits<ForwardIt>::value_type>::value,
+                  "shape_size expects 2 forward iterators as inputs");
+
```
Comment (lazarevevgeny) [inline ngraph/core/include/ngraph/shape.hpp:64]: Should it throw an exception if the dimension is dynamic?

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp (hunk 3 above)
Comment (lazarevevgeny) [reply ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:142]: Please, add code comments so we will not loose this information.

File: ngraph/core/src/op/max_pool.cpp [hunk 8]
```diff
@@ -160,6 +160,102 @@ bool op::v1::MaxPool::has_evaluate() const {
 
 // ------------------------------ V8 ------------------------------
 
+namespace maxpool_v8 {
+
+template <element::Type_t Values, element::Type_t Indices>
+inline bool evaluate(const HostTensorPtr& data,
```
Comment (lazarevevgeny) [inline ngraph/core/src/op/max_pool.cpp:166]: @tomdol, please, tell us what is the binary increase size caused by this evaluate?

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp (hunk 2 above)
Comment (lazarevevgeny) [inline ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:152]: Need an assert that data_shape.size() >= 2. Otherwise we may have an infinite loop

Comment (lazarevevgeny) [reply ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:152]: Or even better solution is to change the code the following way:
```cpp
for (size_t dim = 0; dim + 2 < data_shape.size(); ++dim) {
```

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp [hunk 9]
```diff
@@ -110,6 +110,340 @@ void max_pool(const T* arg,
...
+                                 const Shape& pads_end) {
+    kernel_position[2] += kernel_strides[2];
+
+    if (kernel_position[2] + (kernel[2] - 1) * kernel_dilations[2] >= data_shape[4] + pads_end[2]) {
+        kernel_position[2] = 0 - pads_begin[2];
+        kernel_position[1] += kernel_strides[1];
+        if (kernel_position[1] + (kernel[1] - 1) * kernel_dilations[1] >= data_shape[3] + pads_end[1]) {
+            kernel_position[1] = 0 - pads_begin[1];
+            kernel_position[0] += kernel_strides[0];
+        }
```
Comment (lazarevevgeny) [inline ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:199]: You can simply call `next_kernel_position_2D` instead

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp [hunk 10]
```diff
@@ -110,6 +110,340 @@ void max_pool(const T* arg,
...
+                 const size_t kernel_stride,
+                 const size_t kernel_dilation,
+                 const size_t pads_begin,
+                 const size_t pads_end,
+                 const size_t indices_offset) {
+    int kernel_position = 0 - pads_begin;
+    // select max elem and its index for each "placeholder" in the out buffer (pointed to by out_idx)
+    for (size_t out_idx = 0; out_idx < out_elems; ++out_idx) {
+        Values_t max_elem = std::numeric_limits<Values_t>::lowest();
+        Indices_t max_elem_idx = Indices_t{0};
```
Comment (lazarevevgeny) [inline ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:222]: Minor optimization. You can initialize the `max_elem_idx` with the index of the element in the top left corner of the kernel and initialize `max_elem` with the value of that element. After that you can iterate `kernel_elem` starting from 1 in line 223 instead of 0.

Comment (lazarevevgeny) [reply ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:222]: The same comment is applicable for other `max_pool_Xd` functions.

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp [hunk 11]
```diff
@@ -110,6 +110,340 @@ void max_pool(const T* arg,
...
+            const size_t kernel_elem_offset = kernel_elem * kernel_dilation;
+            if (kernel_position + kernel_elem_offset < 0 || kernel_position + kernel_elem_offset >= data_elems) {
+                // don't process the padding elements
+                continue;
+            } else {
+                if (data[kernel_position + kernel_elem_offset] > max_elem) {
+                    max_elem = data[kernel_position + kernel_elem_offset];
+                    max_elem_idx = kernel_position + kernel_elem_offset;
+                }
+            }
```
Comment (lazarevevgeny) [inline ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:233]: The branch of `if` statement which will likely be true should not be in `else` branch. This can result in severe performance degradation due to branch prediction in the CPU. So, I would insist on refactoring this code to:

```cpp
            if (kernel_position + kernel_elem_offset >= 0 && kernel_position + kernel_elem_offset < data_elems && data[kernel_position + kernel_elem_offset] > max_elem) {
                max_elem = data[kernel_position + kernel_elem_offset];
                max_elem_idx = kernel_position + kernel_elem_offset;
            }
```

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp (hunk 10 above)
Comment (lazarevevgeny) [reply ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:222]: I would also recommend to get rid of `max_elem` variable at all. You can use just `max_elem_idx` variable like this: `data[max_elem_idx]` instead of updating two variables each time you find a new max element. CPU cache will keep the value of `max_elem_idx` in some register so accessing it will be very fast.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]
From the math perspective changes looks good, but there are a number of simple optimizations which should be applied.

File: ngraph/core/src/op/max_pool.cpp (hunk 6 above)
Comment (lazarevevgeny) [reply ngraph/core/src/op/max_pool.cpp:342]: Agree with @ilyachur. The fallback to ref implementation is supported on CPU only and MaxPoll-8 will be supported on CPU. So it doesn't make sense to have it in the evaluate.

Comment (tomdol) [reply ngraph/core/src/op/max_pool.cpp:342]: @ilyachur @lazarevevgeny does that mean I should move it to evaluates_map or do you mean something else?

File: ngraph/core/include/ngraph/shape.hpp (hunk 7 above)
Comment (tomdol) [reply ngraph/core/include/ngraph/shape.hpp:64]: In fact it will not compile. If you use this function with something that stores `Dimension` objects, the `static_assert` will fire because dimension is not an arithmetic type. This overload only works with simple PODs like `int` or `size_t` and that's why I added this compile-time check. I will change the error message though because it's not accurate right now.

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp (hunk 9 above)
Comment (tomdol) [reply ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:199]: If I call the 2D version I will get a 2D Coord object in return. I will have to merge it into the currently processed 3D version. This requires copying or moving the underlying integers which be less performant than having this slightly duplicated code.

File: ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp (hunk 10 above)
Comment (tomdol) [reply ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp:222]: Removing the `max_elem` variable complicates the code a bit because I have to find the first element that doesn't belong to the padding area. Only then I can use its value to initialize `data[max_elem_idx]` for the first time. I can initialize the `max_elem` this way: `Values_t max_elem = kernel_position < 0 ? std::numeric_limits<Values_t>::lowest() : data[kernel_position];` and we'll save one iteration of the inner loop but this will add an if-statement so I'm not sure if it's worth it. 
I'd rather leave it as is to keep the code more readable.

File: ngraph/core/src/op/max_pool.cpp (hunk 6 above)
Comment (ilyachur) [reply ngraph/core/src/op/max_pool.cpp:342]: Yes, looks like at the current moment we don't need to have evaluate method for this operation, but this implementation has a lot of template code and can have significant impact on binary size.

File: ngraph/core/src/op/max_pool.cpp (hunk 8 above)
Comment (tomdol) [reply ngraph/core/src/op/max_pool.cpp:166]: @lazarevevgeny the increase is 72kB in Release mode

**Review by lazarevevgeny — APPROVED** [review]

**Review by mitruska — APPROVED** [review]
